### PR TITLE
Pass correct `linkage_name` attribute to DIBuilder

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
New file created `dppy_debuginfo.py` with the `DPPYDIBuilder` class.
`DPPYDIBuilder` modifies `mark_subprogram` method and passes  correct name to `linkage_name` attribute.

New `ModifiedLower` class contains the constructor needed to create an instance of `DPPYDIBuilder` class.

Tests:
- [x] IR contains `DISubprogram` with function `linkageName` without ID
- [x] IR contains `DICompileUnit` with `DW_LANG_C_plus_plus`
- [x] IR contains `DICompileUnit` with `producer = numba-dppy`

Changes:
- Removed `ModifiedLower`. `DPPYLower` adjusts standard `Lower`.

### CODE REVIEW

File: numba_dppy/dppy_debuginfo.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from numba.core.debuginfo import DIBuilder
+
+
+class DPPYDIBuilder(DIBuilder):
+    def mark_subprogram(self, function, name, loc):
+        if isinstance(name, tuple):
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_debuginfo.py:6]: See below. Tuple could be replaced by using `DPPYDIBuilder` member for `mangled_qualname`.

**Review by PokhodenkoSA — CHANGES_REQUESTED** [review]

File: numba_dppy/dppy_lowerer.py [hunk 2]
```diff
@@ -1207,9 +1208,73 @@ def get_slots_members(obj):
...
+        self.blkmap = {}
+        self.pending_phis = {}
+        self.varmap = {}
+        self.firstblk = min(self.blocks.keys())
+        self.loc = -1
+
+        # Specializes the target context as seen inside the Lowerer
+        # This adds:
+        #  - environment: the python execution environment
+        self.context = context.subtarget(environment=self.env, fndesc=self.fndesc)
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply numba_dppy/dppy_lowerer.py:1247]: I tried this suggestion earlier, but then we create two IR modules:
https://github.com/IntelPython/numba-dppy/pull/434/files#diff-6c8257f1d30d71128472a57b8a7e3e1d791915b259eb6116988b3465ab3f68fbR1231
I do not know if this is correct but breakpoints by function name work.
image
image

File: numba_dppy/dppy_lowerer.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import linecache
 import os
 import sys
+from numba import core
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply numba_dppy/dppy_lowerer.py:21]: Yes. I will delete this line.

File: numba_dppy/dppy_debuginfo.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from numba.core.debuginfo import DIBuilder
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply numba_dppy/dppy_debuginfo.py:1]: I will add license.

File: numba_dppy/dppy_lowerer.py [hunk 5]
```diff
@@ -1207,9 +1208,73 @@ def get_slots_members(obj):
...
+        self.loc = -1
+
+        # Specializes the target context as seen inside the Lowerer
+        # This adds:
+        #  - environment: the python execution environment
+        self.context = context.subtarget(environment=self.env, fndesc=self.fndesc)
+
+        # Debuginfo
+        if self.context.enable_debuginfo:
+            self.debuginfo = DPPYDIBuilder(
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply numba_dppy/dppy_lowerer.py:1251]: Is it correct to change the signature of a function in a child class? Should not the signature match the parent class?
https://github.com/IntelPython/numba/blob/release0.53-patched/numba/core/debuginfo.py#L141

[the remark this answers is not in the document]
Comment (PokhodenkoSA) [reply numba_dppy/dppy_lowerer.py:1222]: Yes, if the child class support the additional parameter.

File: numba_dppy/dppy_debuginfo.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from numba.core.debuginfo import DIBuilder
+
+
+class DPPYDIBuilder(DIBuilder):
+    def __init__(self, module, filepath, mangled_qualname):
+        DIBuilder.__init__(self, module, filepath)
+        self.linkage_name = mangled_qualname
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_debuginfo.py:21]:
```suggestion
    def __init__(self, module, filepath, linkage_name ):
        DIBuilder.__init__(self, module, filepath)
        self.linkage_name = linkage_name 
```

Comment (PokhodenkoSA) [reply numba_dppy/dppy_debuginfo.py:21]: no need for new name

File: numba_dppy/dppy_lowerer.py [hunk 7]
```diff
@@ -1207,9 +1207,32 @@ def get_slots_members(obj):
...
-class DPPYLower(Lower):
+from numba.core.funcdesc import qualifying_prefix, default_mangler
+from numba_dppy.dppy_debuginfo import DPPYDIBuilder
+from numba.core import types, ir, debuginfo, funcdesc, config
+
+
+class ModifiedLower(Lower):
     def __init__(self, context, library, fndesc, func_ir, metadata=None):
         Lower.__init__(self, context, library, fndesc, func_ir, metadata)
+        self.qualprefix = qualifying_prefix(fndesc.modname, fndesc.qualname)
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_lowerer.py:1218]: Used anywhere?

**Review by PokhodenkoSA — APPROVED** [review]
Added tests and code cleanup.

**Pull request merged**